modules/contact/include/utils/MortarContactUtils.h

```diff
@@ -136,6 +160,202 @@ communicateVelocities(std::unordered_map<const DofObject *, T> & dof_map,
+  std::unordered_map<processor_id_type, std::vector<Datum>> push_data;
+
+  for (auto & pr : dof_map_adr2t)
+  {
+    const auto * const dof_object = pr.first;
+    const auto proc_id = dof_object->processor_id();
+    if (proc_id == our_proc_id)
+      continue;
+
+    push_data[proc_id].push_back(std::make_pair(dof_object->id(), std::move(pr.second)));
+  }
+
+  const auto & lm_mesh = mesh.getMesh();
+  std::unordered_map<processor_id_type, std::vector<const DofObject *>>
+      pid_to_dof_object_for_sending_back;
+
+  auto action_functor =
+      [nodal,
+       our_proc_id,
+       &lm_mesh,
+       &dof_map_adr2t,
+       &pid_to_dof_object_for_sending_back,
+       send_data_back](const processor_id_type pid, const std::vector<Datum> & sent_data)
+  {
+    mooseAssert(pid != our_proc_id, "We do not send messages to ourself here");
+    libmesh_ignore(our_proc_id);
+
+    for (auto & pr : sent_data)
+    {
+      const auto dof_id = pr.first;
+      const auto * const dof_object =
+          nodal ? static_cast<const DofObject *>(lm_mesh.node_ptr(dof_id))
+                : static_cast<const DofObject *>(lm_mesh.elem_ptr(dof_id));
+      mooseAssert(dof_object, "This should be non-null");
+
+      if (send_data_back)
+        pid_to_dof_object_for_sending_back[pid].push_back(dof_object);
+
+      for (const auto i : make_range(3))
+        for (const auto j : make_range(3))
+          dof_map_adr2t[dof_object](i, j) += pr.second(i, j);
+    }
+  };
+
+  TIMPI::push_parallel_vector_data(communicator, push_data, action_functor);
+
+  // Now send data back if requested
+  if (!send_data_back)
+    return;
+
+  std::unordered_map<processor_id_type, std::vector<Datum>> push_back_data;
+
+  for (const auto & [pid, dof_objects] : pid_to_dof_object_for_sending_back)
+  {
+    auto & pid_send_data = push_back_data[pid];
+    pid_send_data.reserve(dof_objects.size());
+    for (const DofObject * const dof_object : dof_objects)
+    {
+      const auto & r2t = libmesh_map_find(dof_map_adr2t, dof_object);
+      pid_send_data.push_back({dof_object->id(), r2t});
+    }
+  }
+
+  auto sent_back_action_functor =
+      [nodal, our_proc_id, &lm_mesh, &dof_map_adr2t](const processor_id_type libmesh_dbg_var(pid),
+                                                     const std::vector<Datum> & sent_data)
+  {
+    mooseAssert(pid != our_proc_id, "We do not send messages to ourself here");
+    libmesh_ignore(our_proc_id);
+
+    for (auto & [dof_id, r2t_sent] : sent_data)
+    {
+      const auto * const dof_object =
+          nodal ? static_cast<const DofObject *>(lm_mesh.node_ptr(dof_id))
+                : static_cast<const DofObject *>(lm_mesh.elem_ptr(dof_id));
+      mooseAssert(dof_object, "This should be non-null");
+      auto & r2t = dof_map_adr2t[dof_object];
+      r2t = r2t_sent;
+    }
+  };
+
+  TIMPI::push_parallel_vector_data(communicator, push_back_data, sent_back_action_functor);
+}
+
+template <typename T>
```

Comment:
You have a template parameter here, but the name implies it should always be `ADReal`?

Reply:
I have tested `ADReal` and `ADRealVectorValue` for that method. Need to come up with a name that would make sense.

Reply:
Maybe `communicateADRealObject`.